### ISSUE TITLE
ModConfig syncing on join fails if serialized json exceeds ModPacket limit of 65535

### ISSUE THREAD

**Issue by direwolf420** [tModLoader/tModLoader#779]
Possible solution: Split up ModConfig json

**Comment by Chicken-Bones** [comment]
Arguably we just need some packet splitting logic. But also... You have a 65kb config? Compressed?

**Comment by direwolf420** [comment]
[RecipeTweaker2_Config.zip](https://github.com/tModLoader/tModLoader/files/4139065/RecipeTweaker2_Config.zip)

**Comment by Rartrin** [comment]
What is the reason for the length being a ushort? If it is just a matter of keeping packets small because most don't have a lot of data, then we could just use 7bit encoded ints for length or something similar.

**Comment by Chicken-Bones** [comment]
Main reason is that packet length is pretty high level in the protocol. And we'd want to keep at least the version negotiation on the Vanilla spec. 

We could switch to 7bit enc once we know both sides are using tModLoader though yes.

**Comment by direwolf420** [comment]
> Main reason is that packet length is pretty high level in the protocol. And we'd want to keep at least the version negotiation on the Vanilla spec.
> 
> We could switch to 7bit enc once we know both sides are using tModLoader though yes.

Since vanilla compatibility is being ditched in 1.4.4, is this possible then?

**Comment by Chicken-Bones** [comment]
After investigating further, this would require allocating larger send and receive buffers for every client connected to a server, consuming a potentially large amount of memory. Dynamically sizing the buffers would require a significant rewrite (or a lot of hacky bodge). 

In addition, a fixed packet length offers good protection against stalling the main thread due to excessive time spent serializing/deserializing/copying, and against extra large packets crashing the process. If the user is on a low-bandwidth connection (say, 10KB/s), large packets also offer no progress reporting mechanism so even if the process isn't hung, they could be stuck in the connecting state for a long time.

All in all, configs need to be sent in 'chunks', just like mods are.

**Comment by zzp198** [comment]
Chunks still need to be concatenated in the end, perhaps the memory used and direct allocation are not much different, and there is no efficient splicing method like "ByteBuilder".

**Comment by zzp198** [comment]
Yes, when I try to send packets in segments at once, both asynchronous and threaded methods will inevitably cause stuttering. Perhaps only transmitting 65535 in a single tick is the best choice.
